File: Algorithm/blog/o5/kmp.php
<?php 

/*

kmp算法有两个关键点：1.计算模式字符串的部分匹配表（这时候，自己跟自己比较）2.匹配主串时候，主串字符只遍历一遍，匹配时候，根据模式串的部分匹配表计算模式串应该移动的位置。kmp算法时间复杂度为O(m+n);
 */

function kmp_next($string){
    $length = strlen($string);//获取字符串长度
    $next[0] =0;
    $j =0;
    $i = 1;
    while($i<$length){
        if($string{$j} == $string{$i}){
            $j++;
            $next[$i]= $j;
            $i++;
        }else if($j == 0){
            $next[$i] =$j;
            $i++;
        }else{
            $j=$next[$j];
        }
    }
    return $next;
}

function kmp($text,$mode){
    $t_length = strlen($text);
    $m_length = strlen($mode);
    if($t_length < $m_length){
        return -1;
    }
    $arr = kmp_next($mode);
    $j=0;$i=0;
    while($i<$t_length){
        if($text{$i}==$mode{$j}){
            if($j < $m_length-1){
                $j++;$i++;
            }else{
                return $i-$m_length+1;
            }
        }else if($j==0){
            $i++;
        }else{
            $j=$arr[$j-1];
        }
    }
    return -1;
}


$string = 'BBC ABCDAB ABCDABCDABDE';
$mode = 'ABCDABD';

$key = kmp($string,$mode);
var_dump(kmp_next($mode));
var_dump($key);
var_dump(substr($string,$key,strlen($mode)));    

File: Algorithm/search/code/KMP.java
public class KMP {
    public static int KMPSearch(String txt, String pat, int[] next) {
        int M = txt.length();
        int N = pat.length();
        int i = 0;
        int j = 0;
        while (i < M && j < N) {
            if (j == -1 || txt.charAt(i) == pat.charAt(j)) {
                i++;
                j++;
            } else {
                j = next[j];
            }
        }
        if (j == N)
            return i - j;
        else
            return -1;
    }
    public static void getNext(String pat, int[] next) {
        int N = pat.length();
        next[0] = -1;
        int k = -1;
        int j = 0;
        while (j < N - 1) {
            if (k == -1 || pat.charAt(j) == pat.charAt(k)) {
                ++k;
                ++j;
                next[j] = k;
            } else
                k = next[k];
        }
    }
    public static void main(String[] args) {
        String txt = "BBC ABCDAB CDABABCDABCDABDE";
        String pat = "ABCDABD";
        int[] next = new int[pat.length()];
        getNext(pat, next);
        System.out.println(KMPSearch(txt, pat, next));
    }
}